msm/src/mvlookup.rs

```diff
@@ -0,0 +1,403 @@
+// main.rs. It should be moved to the test API when main.rs is replaced with
+// real production code.
+impl<F: FftField> LookupWitness<F> {
+    /// Generate a random number of correct lookups in the table RangeCheck16
+    pub fn random(domain: EvaluationDomains<F>) -> Self {
+        let mut rng = thread_rng();
+        // TODO: generate more random f
+        let table_size: u64 = rng.gen_range(1..domain.d1.size);
+        let table_id = rng.gen_range(1..1000);
+        // Build a table of value we can look up
+        let t: Vec<u64> = {
+            // Generate distinct values to avoid to have to handle the
+            // normalized multiplicity polynomial
+            let mut n: Vec<u64> = (1..(table_size * 100)).collect();
+            n.shuffle(&mut rng);
+            n[0..table_size as usize].to_vec()
+        };
+        // permutation argument
+        let f = {
+            let mut f = t.clone();
+            f.shuffle(&mut rng);
+            f
+        };
+        let dummy_value = F::rand(&mut rng);
+        let repeated_dummy_value: Vec<F> = {
+            let r: Vec<F> = iter::repeat(dummy_value)
+                .take((domain.d1.size - table_size) as usize)
+                .collect();
+            r
+        };
+        let t_evals = {
+            let mut table = Vec::with_capacity(domain.d1.size as usize);
+            table.extend(t.iter().map(|v| Lookup {
+                table_id: LookupTable::Custom(table_id),
+                numerator: -F::one(),
+                value: vec![F::from(*v)],
+            }));
+            table.extend(
+                repeated_dummy_value
+                    .iter()
+                    .map(|v| Lookup {
+                        table_id: LookupTable::Custom(table_id),
+                        numerator: -F::one(),
+                        value: vec![*v],
+                    })
+                    .collect::<Vec<Lookup<F>>>(),
+            );
+            table
+        };
+        let f_evals: Vec<Lookup<F>> = {
+            let mut table = Vec::with_capacity(domain.d1.size as usize);
+            table.extend(f.iter().map(|v| Lookup {
+                table_id: LookupTable::Custom(table_id),
+                numerator: F::one(),
+                value: vec![F::from(*v)],
+            }));
+            table.extend(
+                repeated_dummy_value
+                    .iter()
+                    .map(|v| Lookup {
+                        table_id: LookupTable::Custom(table_id),
+                        numerator: F::one(),
+                        value: vec![*v],
+                    })
+                    .collect::<Vec<Lookup<F>>>(),
+            );
+            table
+        };
+        let m = (0..domain.d1.size).map(|_| F::one()).collect();
+        LookupWitness {
+            f: vec![f_evals],
+            t: t_evals,
+            m,
+        }
+    }
+}
+
+/// Represents the proof of the lookup argument
+/// It is parametrized by the type `T` which can be either:
```

Comment:
What do you think about using explicit fields instead? `m_com, h_com, sum_com, m_eval, h_eval, sum_eval`? Or two structs even, `LookupProofComs` and `LookupProofEvals`?

Reply:
I'm in favor of keeping low the number of different structures, and hide the differences in the type. Maybe we can rename the structure.